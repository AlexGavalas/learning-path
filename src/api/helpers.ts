import axios from 'axios';

export const fetchFileFromStorage = async (
    filename: string,
): Promise<string> => {
    const fileServerUrl = String(process.env.PUBLIC_FILE_SERVER_URL);

    const response = await axios.get(`${fileServerUrl}/${filename}`, {
        responseType: 'text',
    });

    return response.data as string;
};
