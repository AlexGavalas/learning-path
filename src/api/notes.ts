import type { AstroComponentFactory } from 'astro/dist/runtime/server';
import { getEntry, render } from 'astro:content';
import flow from 'lodash/fp/flow';
import groupBy from 'lodash/fp/groupBy';
import mapValues from 'lodash/fp/mapValues';

import type { BlogArticleFrontmatter } from '~types/blog';
import type { Note, NoteFrontmatter } from '~types/notes';
import type { SummaryFrontmatter } from '~types/summaries';

import { getNoteMetadata, searchNotes } from './notes-db';

type SearchNotesRpcResponse = {
    title: string;
    line: string;
}[];

type Lines = Record<string, string[]>;

const groupByTitle = flow(
    groupBy<SearchNotesRpcResponse[number]>(({ title }) => title),
    mapValues((notes) => notes.map(({ line }) => line)),
);

export const fetchNotes = async (
    query: string,
): Promise<{
    lines: Lines;
    notes: Note[];
}> => {
    const allNotes = await getNoteMetadata();

    if (query.length > 0) {
        const rows = await searchNotes(query);

        const lines = groupByTitle(rows);

        const noteTitles = new Set(rows.map(({ title }) => title));

        const filteredNotes = allNotes.filter(({ title }) =>
            noteTitles.has(title),
        );

        return {
            lines,
            notes: filteredNotes,
        };
    }

    return {
        lines: {},
        notes: allNotes,
    };
};

export const getNoteData = async ({
    collection,
    slug,
}: {
    collection: 'blog' | 'notes' | 'summaries';
    slug: string;
}): Promise<{
    content: { Content: AstroComponentFactory };
    frontmatter: NoteFrontmatter | SummaryFrontmatter | BlogArticleFrontmatter;
} | null> => {
    const note = await getEntry(collection, slug);

    if (!note) {
        return null;
    }

    return {
        content: await render(note),
        frontmatter: note.data,
    };
};
