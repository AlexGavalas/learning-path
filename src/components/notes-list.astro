---
import List from '~components/list.astro';
import ListItem from '~components/list-item.astro';
import FormattedDate from '~components/formatted-date.astro';
import type { Note } from '~types/notes.types';

type _Note = Pick<Note, 'filename' | 'updated' | 'title'>;

export type Props = {
    notes: _Note[];
    lines?: Record<string, string[]>;
    timeZone: string;
    baseUrl: string;
};

const { notes, lines, timeZone, baseUrl } = Astro.props;
const hasNotes = notes.length > 0;
---

{
    hasNotes ? (
        <ul class="list-none divide-x-0 divide-y-2 divide-solid divide-zinc-300 p-0 dark:divide-zinc-800">
            {notes.map(({ filename, updated, title }) => (
                <li>
                    <a
                        href={`/${baseUrl}/${filename}`}
                        class="hover:no-underline"
                    >
                        <div class="group flex cursor-pointer items-center justify-between p-2 text-black hover:bg-gray-100 dark:text-white dark:hover:bg-neutral-800">
                            <div class="flex gap-2">
                                <FormattedDate
                                    dateString={updated}
                                    timeZone={timeZone}
                                    format="dd/MM/yy"
                                />
                                <span class="text-light-primary hover:no-underline dark:text-dark-primary">
                                    {title}
                                </span>
                            </div>
                            <span class="opacity-0 group-hover:animate-pulse">
                                &#x21dd;
                            </span>
                        </div>
                    </a>
                    {lines?.[title] && (
                        <List>
                            {lines[title].map((line) => (
                                <ListItem key={line}>{line}</ListItem>
                            ))}
                        </List>
                    )}
                </li>
            ))}
        </ul>
    ) : (
        <p class="text-center">No notes found</p>
    )
}
