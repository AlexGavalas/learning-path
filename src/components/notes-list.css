.link {
    display: grid;
    grid-template-columns: auto 1fr auto;
    gap: 8px;
    padding: 8px;
    @apply hover:bg-[#e2e2e2] dark:hover:bg-neutral-800;
}

.link:hover::after {
    @apply animate-pulse;
    content: '⇝';
}

.text {
    @apply text-light-primary dark:text-dark-primary;
}
