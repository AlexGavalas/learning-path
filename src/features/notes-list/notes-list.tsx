import Link from 'next/link';

import { FormattedDate } from '~components/formatted-date';
import { List, ListItem } from '~components/list';
import { type Note } from '~types/notes.types';

type NotesListProps = {
    notes: Note[];
    lines: Record<string, string[]>;
};

export const NotesList = ({ notes, lines }: NotesListProps) => {
    if (!notes.length) {
        return <p className="text-center">No notes found</p>;
    }

    return (
        <ul className="list-none divide-x-0 divide-y-2 divide-solid divide-zinc-300 p-0 dark:divide-zinc-800">
            {notes.map(({ filename, updated, title }) => (
                <li key={filename}>
                    <Link
                        href={`/notes/${filename}`}
                        className="hover:no-underline"
                        role="link"
                    >
                        <div className="group flex cursor-pointer items-center justify-between p-2 text-black hover:bg-gray-100 dark:text-white dark:hover:bg-neutral-800">
                            <div className="flex flex-col-reverse gap-2 sm:flex-row">
                                <FormattedDate
                                    dateString={updated}
                                    format="dd/MM/yy"
                                />
                                <span className="text-light-primary hover:no-underline dark:text-dark-primary">
                                    {title}
                                </span>
                            </div>
                            <span className="opacity-0 group-hover:animate-pulse">
                                &#x21dd;
                            </span>
                        </div>
                    </Link>
                    {lines[title] && (
                        <List>
                            {lines[title].map((line) => (
                                <ListItem key={line}>{line}</ListItem>
                            ))}
                        </List>
                    )}
                </li>
            ))}
        </ul>
    );
};
