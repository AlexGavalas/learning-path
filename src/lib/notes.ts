import matter from 'gray-matter';
import { serialize } from 'next-mdx-remote/serialize';
import rehypeAutolinkHeadings from 'rehype-autolink-headings';
import rehypeExternalLinks from 'rehype-external-links';
import rehypeSlug from 'rehype-slug';

import { supabase } from '~lib/supabase';
import { type NoteMDX } from '~types/notes.types';

export const getAllNoteIds = async () => {
    const { data: fileNames, error } = await supabase.storage
        .from('notes_test_1')
        .list();

    if (error) {
        throw error;
    }

    return fileNames.map((file) => {
        return {
            params: {
                id: file.name.replace(/\.mdx$/, ''),
            },
        };
    });
};

export const getNoteData = async (filename: string): Promise<NoteMDX> => {
    const filePath = `${filename}.mdx`;

    const { data: fileContents, error } = await supabase.storage
        .from('notes_test_1')
        .download(filePath);

    if (error) {
        throw error;
    }

    const matterResult = matter(await fileContents.text());

    const mdxSource = await serialize(matterResult.content, {
        mdxOptions: {
            rehypePlugins: [
                [rehypeExternalLinks, { target: '_blank' }],
                [rehypeSlug],
                [rehypeAutolinkHeadings, { behavior: 'wrap' }],
            ],
        },
    });

    // Can't type gray matter result yet
    return {
        mdxSource,
        ...matterResult.data,
    } as NoteMDX;
};
