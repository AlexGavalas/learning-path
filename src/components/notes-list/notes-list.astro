---
import List from '~components/list/list.astro';
import ListItem from '~components/list/list-item.astro';
import FormattedDate from '~components/formatted-date.astro';
import type { Note } from '~types/notes.types';
import * as cn from './notes-list.css';

interface Props {
    notes: readonly Note[];
    lines?: Record<string, string[]>;
    baseUrl: string;
}

const { notes, lines, baseUrl } = Astro.props;
const hasNotes = notes.length > 0;
---

{
    hasNotes ? (
        <ul>
            {notes.map(({ filename, updated, title }, index) => {
                const dateId = `${index}d`;
                const textId = `${index}t`;

                return (
                    <li class={cn.item}>
                        <a
                            aria-labelledby={`${textId} ${dateId}`}
                            class={cn.link}
                            href={`/${baseUrl}/${filename}`}
                        >
                            <FormattedDate
                                dateString={updated}
                                format="DD/MM/YY"
                                id={dateId}
                                label="Last updated at"
                            />
                            <span
                                aria-label={`Notes about ${title}`}
                                class={cn.text}
                                id={textId}
                            >
                                {title}
                            </span>
                        </a>
                        {lines?.[title] && (
                            <List>
                                {lines[title].map((line) => (
                                    <ListItem key={line}>{line}</ListItem>
                                ))}
                            </List>
                        )}
                    </li>
                );
            })}
        </ul>
    ) : (
        <p class={cn.noNotes}>No notes found</p>
    )
}
