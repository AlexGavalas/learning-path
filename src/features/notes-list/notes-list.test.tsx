import { axe } from 'jest-axe';
import { type ReactElement, cloneElement } from 'react';

import { getTimeZone, renderWithUser, screen } from '~test/helpers';
import { type Note } from '~types/notes.types';

import { NotesList } from './notes-list';

jest.mock(
    'next/link',
    () =>
        ({
            children,
            prefetch,
            ...rest
        }: {
            prefetch: boolean;
            children: ReactElement;
        }) =>
            cloneElement(children, { prefetch: prefetch?.toString(), ...rest }),
);

const timeZone = getTimeZone();

describe('<NotesList />', () => {
    describe('when there are no notes', () => {
        const NO_NOTES: Note[] = [];
        const NO_LINES = {};

        it('renders', () => {
            const { container } = renderWithUser(
                <NotesList
                    baseUrl="notes"
                    notes={NO_NOTES}
                    lines={NO_LINES}
                    timeZone={timeZone}
                />,
            );

            expect(container).toMatchSnapshot();
        });

        it('is accessible', async () => {
            const { container } = renderWithUser(
                <NotesList
                    baseUrl="notes"
                    notes={NO_NOTES}
                    lines={NO_LINES}
                    timeZone={timeZone}
                />,
            );

            const a11yResults = await axe(container);

            expect(a11yResults).toHaveNoViolations();
        });
    });

    describe('when there are notes', () => {
        const NOTES: Note[] = [
            {
                created: '2022-01-01',
                id: 1,
                filename: 'filename',
                line: 'line',
                title: 'title',
                updated: '2022-01-01',
            },
        ];

        const NO_LINES = {};

        it('renders', () => {
            const { container } = renderWithUser(
                <NotesList
                    baseUrl="notes"
                    notes={NOTES}
                    lines={NO_LINES}
                    timeZone={timeZone}
                />,
            );

            expect(container).toMatchSnapshot();
        });

        it('is accessible', async () => {
            const { container } = renderWithUser(
                <NotesList
                    baseUrl="notes"
                    notes={NOTES}
                    lines={NO_LINES}
                    timeZone={timeZone}
                />,
            );

            const a11yResults = await axe(container);

            expect(a11yResults).toHaveNoViolations();
        });

        it('renders item in a link', () => {
            renderWithUser(
                <NotesList
                    baseUrl="notes"
                    notes={NOTES}
                    lines={NO_LINES}
                    timeZone={timeZone}
                />,
            );

            const link = screen.getByRole('link', {
                name: new RegExp(NOTES[0].title),
            });

            expect(link).toBeInTheDocument();
        });

        describe('when there are lines', () => {
            const LINES = {
                [NOTES[0].title]: ['line'],
            };

            it('renders', () => {
                const { container } = renderWithUser(
                    <NotesList
                        baseUrl="notes"
                        notes={NOTES}
                        lines={LINES}
                        timeZone={timeZone}
                    />,
                );

                expect(container).toMatchSnapshot();
            });

            it('is accessible', async () => {
                const { container } = renderWithUser(
                    <NotesList
                        baseUrl="notes"
                        notes={NOTES}
                        lines={LINES}
                        timeZone={timeZone}
                    />,
                );

                const a11yResults = await axe(container);

                expect(a11yResults).toHaveNoViolations();
            });
        });
    });
});
