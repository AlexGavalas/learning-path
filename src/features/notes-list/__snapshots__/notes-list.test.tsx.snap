// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NotesList /> when there are no notes renders 1`] = `
<div>
  <p
    class="text-center"
  >
    No notes found
  </p>
</div>
`;

exports[`<NotesList /> when there are notes renders 1`] = `
<div>
  <ul
    class="list-none divide-x-0 divide-y-2 divide-solid divide-zinc-300 p-0 dark:divide-zinc-800"
  >
    <li>
      <div
        class="hover:no-underline"
        href="/notes/filename"
        role="link"
      >
        <div
          class="flex flex-col-reverse gap-2 sm:flex-row"
        >
          <time
            datetime="2021-12-31T22:00:00.000Z"
          >
            01/01/22
          </time>
          <span
            class="text-light-primary hover:no-underline dark:text-dark-primary"
          >
            title
          </span>
        </div>
        <span
          class="opacity-0 group-hover:animate-pulse"
        >
          ⇝
        </span>
      </div>
    </li>
  </ul>
</div>
`;

exports[`<NotesList /> when there are notes when there are lines renders 1`] = `
<div>
  <ul
    class="list-none divide-x-0 divide-y-2 divide-solid divide-zinc-300 p-0 dark:divide-zinc-800"
  >
    <li>
      <div
        class="hover:no-underline"
        href="/notes/filename"
        role="link"
      >
        <div
          class="flex flex-col-reverse gap-2 sm:flex-row"
        >
          <time
            datetime="2021-12-31T22:00:00.000Z"
          >
            01/01/22
          </time>
          <span
            class="text-light-primary hover:no-underline dark:text-dark-primary"
          >
            title
          </span>
        </div>
        <span
          class="opacity-0 group-hover:animate-pulse"
        >
          ⇝
        </span>
      </div>
      <ul
        class="pb-4"
      >
        <li
          class="list-disc text-[0.9em] marker:text-light-primary dark:marker:text-dark-primary"
        >
          line
        </li>
      </ul>
    </li>
  </ul>
</div>
`;
